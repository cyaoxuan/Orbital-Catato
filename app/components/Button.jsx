import { View } from "react-native";
import { Button } from "react-native-paper";

const PillButton = (props) => {
    return (
        <View style={{ flexDirection: "row" }}>
                <Button style = {{ margin: 10, width: props.width }}
                    disabled={props.disabled}
                    mode={props.mode}
                    onPress={props.onPress}>
                        {props.label}
                </Button>
        </View>
    )
}

export { PillButton };