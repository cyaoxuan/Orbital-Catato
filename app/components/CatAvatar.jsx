import { StyleSheet, TouchableOpacity, View } from "react-native";
import { Avatar } from "react-native-paper";
import { BodyText, TitleText } from "./Text";

// Cat Avatar with circular image and name
// props: photoURL (string), size (image size || number), variant (font size || string), name (string)
const CatAvatar = (props) => {
    return (
        <View style={styles.avatarContainer}>
            <Avatar.Image
                testID="avatar-image"
                source={
                    props.photoURL
                        ? { uri: props.photoURL }
                        : require("../../assets/placeholder.png")
                }
                size={props.size || 100}
            />
            <BodyText variant={props.variant} text={props.name || "Name"} />
        </View>
    );
};

// Cat Avatar with circular image and name, touchable
// props: photoURL (string), size (image size || number), variant (font size || string), name (string), onPress (callback)
const TouchableCatAvatar = (props) => {
    return (
        <TouchableOpacity
            testID="cat-avatar"
            style={styles.avatarContainer}
            activeOpacity={0.7}
            onPress={props.onPress}
        >
            <Avatar.Image
                testID="avatar-image"
                source={
                    props.photoURL
                        ? { uri: props.photoURL }
                        : require("../../assets/placeholder.png")
                }
                size={props.size || 100}
            />
            <BodyText variant={props.variant} text={props.name || "Name"} />
        </TouchableOpacity>
    );
};

export { CatAvatar, TouchableCatAvatar };

const styles = StyleSheet.create({
    avatarContainer: {
        alignItems: "center",
    },
});
