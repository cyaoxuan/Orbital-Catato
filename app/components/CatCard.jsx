import { Card } from "react-native-paper";
import { IconTextField } from "./InfoText";

const CatCard = (props) => {
    return (
        <Card style={{ height: 280, width: 280, margin: 4 }} 
            onPress={() => {}}
            mode="elevated">
            <Card.Cover style={{ height: 140, width: 280, resizeMode: "cover"}} 
                source={props.catImage} 
            />
            <Card.Title title={props.catName} titleVariant="headlineMedium"/>
            <Card.Content style={{ paddingBottom: 0, paddingHorizontal: 4 }}>
                <IconTextField 
                    iconName={props.iconName1} 
                    iconSize={24} 
                    variant="titleLarge"
                    field={props.field1}
                    info={props.info1}
                />
                <IconTextField 
                    iconName={props.iconName2} 
                    iconSize={24} 
                    variant="titleLarge" 
                    field={props.field2}
                    info={props.info2}
                />
            </Card.Content>
        </Card>
    );
};

export { CatCard };