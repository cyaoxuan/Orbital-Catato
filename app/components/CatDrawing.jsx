import { Image } from "react-native";

export const PantingCat = ({ size }) => {
    return (
        <Image
            source={require("../../assets/drawing-cats/cat-drawing-1.png")}
            style={{ height: size, width: size }}
        />
    );
};

export const PopCat = ({ size }) => {
    return (
        <Image
            source={require("../../assets/drawing-cats/cat-drawing-2.png")}
            style={{ height: size, width: size }}
        />
    );
};

export const SmilingGingerCat = ({ size }) => {
    return (
        <Image
            source={require("../../assets/drawing-cats/cat-drawing-3.png")}
            style={{ height: size, width: size }}
        />
    );
};

export const LoadingCat = ({ size }) => {
    return (
        <Image
            source={require("../../assets/drawing-cats/cat-drawing-4.png")}
            style={{ height: size, width: size }}
        />
    );
};

export const MaxwellCat = ({ size }) => {
    return (
        <Image
            source={require("../../assets/drawing-cats/cat-drawing-5.png")}
            style={{ height: size, width: size }}
        />
    );
};

export const SaladSmudgeCat = ({ size }) => {
    return (
        <Image
            source={require("../../assets/drawing-cats/cat-drawing-6.png")}
            style={{ height: size, width: size }}
        />
    );
};

export const StandingCat = ({ size }) => {
    return (
        <Image
            source={require("../../assets/drawing-cats/cat-drawing-7.png")}
            style={{ height: size, width: size }}
        />
    );
};
