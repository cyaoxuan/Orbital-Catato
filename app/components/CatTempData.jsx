// Temp data to showcase static pages

const catViewProps = [{
    iconName1: "information-circle",
    field1: "Status: ",
    iconName2: "location",
    field2: "Last Seen: "
}, {
    iconName1: "time",
    field1: "Last Fed: ",
    iconName2: "location",
    field2: "Last Seen: "
}];

const concernCatCards = [{
    catImage: require("../../assets/temp-cat.jpg"),
    catName: "Kitty",
    iconName1: catViewProps[0].iconName1,
    field1: catViewProps[0].field1,
    info1: "Injured",
    iconName2: catViewProps[0].iconName2,
    field2: catViewProps[0].field2,
    info2: "Engineering"
}, {
    catImage: require("../../assets/temp-cat.jpg"),
    catName: "Skitty",
    iconName1: catViewProps[0].iconName1,
    field1: catViewProps[0].field1,
    info1: "New",
    iconName2: catViewProps[0].iconName2,
    field2: catViewProps[0].field2,
    info2: "Utown"
}, {
    catImage: require("../../assets/temp-cat.jpg"),
    catName: "Mitty",
    iconName1: catViewProps[0].iconName1,
    field1: catViewProps[0].field1,
    info1: "Missing",
    iconName2: catViewProps[0].iconName2,
    field2: catViewProps[0].field2,
    info2: "Science"
}, {
    catImage: require("../../assets/temp-cat.jpg"),
    catName: "Bitty",
    iconName1: catViewProps[0].iconName1,
    field1: catViewProps[0].field1,
    info1: "Injured",
    iconName2: catViewProps[0].iconName2,
    field2: catViewProps[0].field2,
    info2: "Arts"
}];

const fedCatCards = [{
    catImage: require("../../assets/temp-cat.jpg"),
    catName: "Kitty",
    iconName1: catViewProps[1].iconName1,
    field1: catViewProps[1].field1,
    info1: "19/05 12:01",
    iconName2: catViewProps[1].iconName2,
    field2: catViewProps[1].field2,
    info2: "Engineering"
}];

export { concernCatCards, fedCatCards }