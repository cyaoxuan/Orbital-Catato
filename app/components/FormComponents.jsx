import { useState } from "react";
import { View, StyleSheet } from "react-native";
import {
    Button,
    HelperText,
    Menu,
    RadioButton,
    Text,
    TextInput,
} from "react-native-paper";
import { SelectList } from "react-native-dropdown-select-list";
import NumericInput from "react-native-numeric-input";
import RNDateTimePicker from "@react-native-community/datetimepicker";
import { timeOptions } from "../data/DateTimeOptions";

const titleVariant = "titleMedium";
const bodyVariant = "bodyMedium";

// For Location
const DropdownList = (props) => {
    const [selected, setSelected] = useState("");

    return (
        <View style={styles.container}>
            <Text variant={titleVariant}>{props.titleText}</Text>
            <SelectList
                setSelected={
                    props.setSelected
                        ? props.setSelected
                        : (val) => setSelected(val)
                }
                data={props.data || [{ key: "1", value: "value" }]}
                save="value"
            />
        </View>
    );
};

// Text inputs, description reasons, name, features etc
// Required field
const FormInput = (props) => {
    const [value, setValue] = useState("");

    return (
        <View style={styles.container}>
            <TextInput
                style={styles.formInput}
                disabled={props.disabled}
                multiline={props.multiline}
                label={props.label || "Label"}
                placeholder={props.placeholder}
                value={props.value ? props.value : value}
                onChangeText={
                    props.onChangeText ? props.onChangeText : setValue
                }
                error={props.value ? !props.value.trim() : !value.trim()}
            />
            <HelperText
                type="error"
                padding="none"
                visible={props.value ? !props.value.trim() : !value.trim()}
            >
                Error: {props.errorText || "Error"}
            </HelperText>
        </View>
    );
};

// For year
const NumberSpinner = (props) => {
    const [number, setNumber] = useState(new Date().getFullYear());

    return (
        <View style={styles.container}>
            <Text variant={titleVariant}>{props.titleText}</Text>
            <NumericInput
                type="up-down"
                rounded
                editable={false}
                minValue={props.min}
                maxValue={props.max}
                value={props.value ? props.value : number}
                onChange={
                    props.onChange
                        ? props.onChange
                        : (num) => {
                              setNumber(num);
                          }
                }
                totalWidth={120}
                totalHeight={40}
            />
        </View>
    );
};

// For time
const TimeInput = (props) => {
    const today = new Date();
    const [date, setDate] = useState(new Date());
    const [show, setShow] = useState(false);

    const onChange = (event, selectedDate) => {
        const currentDate = selectedDate;
        setShow(false);
        setDate(currentDate);
    };

    return (
        <View style={styles.container}>
            <Text variant={titleVariant}>{props.titleText}</Text>
            <Button
                mode="contained-tonal"
                onPress={props.onPress ? props.onPress : setShow}
                style={{ width: "50%", margin: 4 }}
            >
                Pick Time{" "}
            </Button>
            <Text variant={bodyVariant}>
                Selected Time:{" "}
                {props.displayTime
                    ? props.displayTime.toLocaleTimeString("en-GB", timeOptions)
                    : date.toLocaleTimeString("en-GB", timeOptions)}
            </Text>
            {(props.displayTime ? props.displayTime > today : date > today) && (
                <Text variant={bodyVariant} style={{ color: "crimson" }}>
                    Error: Selected future time! Are you a time traveller?
                </Text>
            )}
            {(props.show ? props.show : show) && (
                <RNDateTimePicker
                    mode="time"
                    display="spinner"
                    is24Hour={true}
                    value={props.value ? props.value : date}
                    onChange={props.onChange ? props.onChange : onChange}
                />
            )}
        </View>
    );
};

// For gender, sterilised, concerns, foster
const TwoRadioInput = (props) => {
    const [selected, setSelected] = useState("First");

    return (
        <View style={styles.container}>
            <Text variant={titleVariant}>{props.titleText}</Text>
            <RadioButton.Group
                value={props.value ? props.value : selected}
                onValueChange={
                    props.onValueChange
                        ? props.onValueChange
                        : (value) => setSelected(value)
                }
            >
                <View style={styles.radioButtonContainer}>
                    <Text variant={bodyVariant}>{props.firstText}</Text>
                    <RadioButton value={props.firstValue || "First"} />
                </View>
                <View style={styles.radioButtonContainer}>
                    <Text variant={bodyVariant}>{props.secondText}</Text>
                    <RadioButton value={props.secondValue || "Second"} />
                </View>
            </RadioButton.Group>
        </View>
    );
};

// To upload photos in form
const UploadPhotos = (props) => {
    const [visible, setVisible] = useState(false);
    const openMenu = () => setVisible(true);
    const closeMenu = () => setVisible(false);

    return (
        <View style={styles.container}>
            <Text variant={titleVariant}>Upload Photos:</Text>
            <View>
                <Menu
                    style={{ marginTop: -50 }}
                    visible={visible}
                    onDismiss={closeMenu}
                    anchor={
                        <Button
                            mode="contained-tonal"
                            icon="camera"
                            onPress={openMenu}
                            style={{ width: "50%", margin: 4 }}
                        >
                            Upload{" "}
                        </Button>
                    }
                >
                    <Menu.Item
                        leadingIcon="camera"
                        onPress={props.cameraOnPress}
                        title="Camera"
                    />
                    <Menu.Item
                        leadingIcon="image"
                        onPress={props.galleryOnPress}
                        title="Gallery"
                    />
                </Menu>
            </View>
            <Text variant={bodyVariant}>
                Image: {props.photoURI ? props.photoURI : ""}
            </Text>
        </View>
    );
};

export {
    DropdownList,
    FormInput,
    NumberSpinner,
    TimeInput,
    TwoRadioInput,
    UploadPhotos,
};

const styles = StyleSheet.create({
    formInput: {
        fontSize: 16,
        width: "100%",
        backgroundColor: "transparent",
        paddingHorizontal: 0,
        alignContent: "center",
    },

    container: {
        width: "90%",
        margin: 4,
    },

    radioButtonContainer: {
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "space-between",
        paddingHorizontal: 4,
    },

    icon: {
        marginHorizontal: 8,
    },
});
