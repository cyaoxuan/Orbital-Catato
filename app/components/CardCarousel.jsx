import { View, FlatList } from "react-native";
import { CatCard } from "../components/CatCard";

const boxWidth = 300;

const Carousel = (props) => {
    return (
        <FlatList style={{ height: 300 }}
                horizontal
                snapToAlignment="center"
                decelerationRate="normal"
                showsHorizontalScrollIndicator={false}
                scrollEventThrottle={10}
                snapToInterval={boxWidth}
                contentOffset={{ x: -120, y: 0 }}
                ItemSeparatorComponent={() => <View style={{width: 20}} />}
                
                data={props.cards}
                renderItem={({item, index}) => {
                    return (
                        <View key={index}> 
                            <CatCard 
                                catImage={item.catImage}
                                catName={item.catName}
                                iconName1={item.iconName1}
                                field1={item.field1}
                                info1={item.info1}
                                iconName2={item.iconName2}
                                field2={item.field2}
                                info2={item.info2}
                            />
                        </View>
                        
                    );
                }}
        />

    );
};

export { Carousel }