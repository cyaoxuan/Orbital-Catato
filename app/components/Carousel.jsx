import { FlatList, View } from "react-native";
import { CatCard } from "./CatCard";
import {
    formatLastFedSimple,
    formatLastSeenSimple,
} from "../utils/formatDetails";

// Function to get info1 field on the card
// @param carouselType: "concern" or "fed"
// @param cat: cat object with concernStatus or lastFedTime
// @return formatted string depending on carouselType or "Unknown" if any fields are missing
function getInfo1(carouselType, cat) {
    if (!cat || !carouselType) {
        return "Unknown";
    }

    if (
        carouselType === "concern" &&
        cat.concernStatus &&
        Object.keys(cat.concernStatus).length !== 0
    ) {
        const concerns = [];
        cat.concernStatus.injured && concerns.push("Injured");
        cat.concernStatus.missing && concerns.push("Missing");
        cat.concernStatus.new && concerns.push("New");

        return concerns.join(", ");
    } else if (carouselType === "unfed" && cat.lastFedTime) {
        return formatLastFedSimple(cat.lastFedTime);
    }
    return "Unknown";
}

// Card Carousel used in Dashboard, implemented using FlatList and snaps to each card
// props: cats (array of cat objects), cardWidth (number), carouselType (string), userRole (object of roles), ...card (props for card)
const CardCarousel = ({
    cats,
    cardWidth,
    carouselType,
    userRole,
    navigation,
    ...card
}) => {
    const spaceBetweenCards = 16;

    return (
        <FlatList
            testID="card-carousel"
            style={{ height: (cardWidth * 6) / 5 }}
            horizontal
            snapToAlignment="center"
            decelerationRate="normal"
            showsHorizontalScrollIndicator={false}
            scrollEventThrottle={4}
            snapToInterval={cardWidth + spaceBetweenCards}
            ItemSeparatorComponent={() => (
                <View style={{ width: spaceBetweenCards }} />
            )}
            data={cats}
            keyExtractor={(item, index) => item.catID}
            renderItem={({ item }) => {
                return (
                    <CatCard
                        name={item.name}
                        photoURL={item.photoURLs ? item.photoURLs[0] : null}
                        cardWidth={cardWidth}
                        profileOnPress={() =>
                            navigation.navigate("catalogue", {
                                screen: "CatProfile",
                                initial: false,
                                params: { catID: item.catID },
                            })
                        }
                        {...card}
                        info1={getInfo1(carouselType, item)}
                        info2={
                            item.locationName && item.locationZone && userRole
                                ? formatLastSeenSimple(
                                      userRole.isCaretaker
                                          ? item.locationName
                                          : item.locationZone,
                                      item.lastSeenTime
                                  )
                                : "Unknown"
                        }
                        showFindLocation={item.lastSeenLocation}
                        locationOnPress={() => {
                            navigation.navigate("Map", {
                                catID: item.catID,
                                location: item.lastSeenLocation,
                            });
                        }}
                        userRole={userRole}
                    />
                );
            }}
        />
    );
};

export { getInfo1, CardCarousel };
