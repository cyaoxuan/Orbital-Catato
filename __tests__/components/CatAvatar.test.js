import { CatAvatar, TouchableCatAvatar } from "../../app/components/CatAvatar";
import { fireEvent, render } from "@testing-library/react-native";

describe("<CatAvatar />", () => {
    it("renders successfully", () => {
        const catAvatar = render(<CatAvatar />);
        expect(catAvatar).toBeDefined();
    });

    it("renders correctly", () => {
        const tree = render(<CatAvatar />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders avatar with correct photoURL", () => {
        const photoURL = require("../../assets/cats/cat-2-1.jpg");
        const { getByTestId } = render(<CatAvatar photoURL={photoURL} />);
        const avatar = getByTestId("avatar-image");
        expect(avatar.props.source.uri).toBe(photoURL);
    });

    it("displays name prop", () => {
        const name = "Kitty";
        const { queryByText } = render(<CatAvatar name={name} />);
        expect(queryByText(name)).not.toBeNull();
    });
});

describe("<TouchableCatAvatar />", () => {
    it("renders successfully", () => {
        const touchableCatAvatar = render(<TouchableCatAvatar />);
        expect(touchableCatAvatar).toBeDefined();
    });

    it("renders correctly", () => {
        const tree = render(<TouchableCatAvatar />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders TouchableOpacity with correct onPress callback", () => {
        const onPressMock = jest.fn();
        const { queryByText } = render(
            <TouchableCatAvatar name="Kitty" onPress={onPressMock} />
        );
        const touchableOpacity = queryByText("Kitty");
        fireEvent.press(touchableOpacity);
        expect(onPressMock).toHaveBeenCalled();
    });

    it("displays name prop", () => {
        const name = "Kitty";
        const { queryByText } = render(<TouchableCatAvatar name={name} />);
        expect(queryByText(name)).not.toBeNull();
    });
});
