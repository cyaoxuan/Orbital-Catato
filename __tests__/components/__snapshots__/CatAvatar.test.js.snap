// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatAvatar /> renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "marginBottom": 4,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#F8F9FD",
          "borderRadius": 50,
          "height": 100,
          "width": 100,
        },
        {
          "backgroundColor": "#F8F9FD",
          "marginBottom": 4,
        },
      ]
    }
  >
    <Image
      accessibilityIgnoresInvertColors={true}
      source={1}
      style={
        {
          "borderRadius": 50,
          "height": 100,
          "width": 100,
        }
      }
      testID="avatar-image"
    />
  </View>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "fontFamily": "System",
          "fontWeight": "400",
          "letterSpacing": 0,
        },
        {
          "writingDirection": "ltr",
        },
        {
          "color": undefined,
          "fontFamily": "Nunito-Medium",
          "textAlign": "center",
        },
      ]
    }
  >
    Name
  </Text>
</View>
`;

exports[`<TouchableCatAvatar /> renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "marginBottom": 4,
      "opacity": 1,
    }
  }
  testID="cat-avatar"
>
  <View
    style={
      [
        {
          "backgroundColor": "#F8F9FD",
          "borderRadius": 50,
          "height": 100,
          "width": 100,
        },
        {
          "backgroundColor": "#F8F9FD",
          "marginBottom": 4,
        },
      ]
    }
  >
    <Image
      accessibilityIgnoresInvertColors={true}
      source={1}
      style={
        {
          "borderRadius": 50,
          "height": 100,
          "width": 100,
        }
      }
      testID="avatar-image"
    />
  </View>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "fontFamily": "System",
          "fontWeight": "400",
          "letterSpacing": 0,
        },
        {
          "writingDirection": "ltr",
        },
        {
          "color": undefined,
          "fontFamily": "Nunito-Medium",
          "textAlign": "center",
        },
      ]
    }
  >
    Name
  </Text>
</View>
`;
